mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -5,821 +5,745 @@
+            .context(WriteError {
+                name: child.name.clone(),
+            })?;
 
-    /// Update checksums and write out MetaData header + index to disk.
-    pub async fn sync_metadata(
-        &mut self,
-        metadata: &mut NexusMetaData,
-    ) -> Result<(), MetaDataError> {
-        metadata.header.index_checksum =
-            MetaDataIndexEntry::checksum(&metadata.index);
-        metadata.header.checksum();
-        self.write_index(&metadata).await
+        Ok(Some(object))
     }
 
-    /// Create a new header + index on "MetaData" partition.
-    pub async fn create_metadata(
-        &mut self,
-    ) -> Result<NexusMetaData, MetaDataError> {
-        let (bdev, _hndl) = self.get_dev().context(NexusChildError {})?;
+    /// Get the most recent entry but do not remove it from the index.
+    pub async fn last(
+        child: &NexusChild,
+    ) -> Result<Option<MetaDataObject>, MetaDataError> {
+        let handle = child.handle().context(HandleError {
+            name: child.name.clone(),
+        })?;
+        let bdev = handle.get_bdev();
+        let block_size = u64::from(bdev.block_len());
 
-        if let Some(partition) = self
-            .probe_label()
+        let mut buf = handle.dma_malloc(block_size).context(ReadAlloc {
+            name: String::from("index"),
+        })?;
+
+        handle
+            .read_at(child.metadata_index_lba * block_size, &mut buf)
             .await
-            .context(ProbeLabelError {})?
-            .partitions
-            .get(0)
-        {
-            if partition.ent_type
-                == GptGuid::from_str(Nexus::METADATA_PARTITION_TYPE_ID).unwrap()
-                && partition.ent_name.name == "MayaMeta"
-            {
-                let mut metadata = NexusMetaData {
-                    header: MetaDataHeader::new(bdev.block_len(), &partition),
-                    index: Vec::new(),
-                };
-                self.sync_metadata(&mut metadata).await?;
-                return Ok(metadata);
-            }
+            .context(ReadError {
+                name: String::from("index"),
+            })?;
+        let index = NexusMetaData::read_index(&buf)?;
+
+        if index.used_entries == 0 {
+            return Ok(None);
         }
 
-        Err(MetaDataError::MissingPartition {})
+        handle
+            .read_at(
+                (index.start_lba + index.current_entry) * block_size,
+                &mut buf,
+            )
+            .await
+            .context(ReadError {
+                name: String::from("object"),
+            })?;
+        let object = NexusMetaData::read_object(&buf)?;
+
+        Ok(Some(object))
     }
 
-    /// Retrieve header + index from "MetaData" partition.
-    pub async fn get_metadata(&self) -> Result<NexusMetaData, MetaDataError> {
-        if let Some(partition) = self
-            .probe_label()
+    /// Return list of the most recent entries but do not modify the index.
```

Comment:
If all entries are being used, do we need to make sure we don't return the oldest entry? Eg, we could've previously attempted to write to it and failed or failed update the index so we can't trust it?

Reply:
We can detect this case if the generation numbers are out of order, but at this stage it would be up to the caller to do that.